### PR TITLE
cli: Remove `allow(clippy::arithmetic_side_effects)`

### PR DESCRIPTION
**Pull request by illia-bobyr**
#### Problem

The rest of the codebase has `clippy::arithmetic_side_effects` enabled.
While CLI is not as critical as the rest of the validator code base, it seems nice to be precise about the arithmetic operations, and clippy helps with that if `arithmetic_side_effects` is enabled.

A blanket disable at the module level is suboptimal.

#### Summary of Changes

As we have a lot of operations already written, that generate warnings under `arithmetic_side_effects`, the first step is to move the lint control into individual modules, that we can then inspect one by one.

Subsequent commits remove `allow(clippy::arithmetic_side_effects)` from all the modules one by one.

---

Not really sure who is the best reviewer for this.
Suggestions are welcome :)

### CODE REVIEW

File: cli/src/checks.rs [hunk 1]
```diff
@@ -84,10 +82,19 @@ pub fn check_account_for_spend_and_fee_with_commitment(
     fee: u64,
     commitment: CommitmentConfig,
 ) -> Result<(), CliError> {
+    let required_balance =
+        balance
+            .checked_add(fee)
+            .ok_or(CliError::InsufficientFundsForSpendAndFee(
```
Comment (brooksprumo) [inline cli/src/checks.rs:88]: Insufficient Funds as an error seems fine. It was strange as first, but I can also see the argument that it would be impossible to have sufficient funds in this overflow case.

File: cli/src/test_utils.rs [hunk 2]
```diff
@@ -46,15 +44,15 @@ pub fn wait_n_slots(rpc_client: &RpcClient, n: u64) -> u64 {
...
+        if new_slot.saturating_sub(slot) > n {
             return new_slot;
         }
     }
 }
 
 pub fn wait_for_next_epoch_plus_n_slots(rpc_client: &RpcClient, n: u64) -> (Epoch, u64) {
     let current_epoch = rpc_client.get_epoch_info().unwrap().epoch;
-    let next_epoch = current_epoch + 1;
+    let next_epoch = current_epoch.saturating_add(1);
```
Comment (brooksprumo) [inline cli/src/test_utils.rs:55]: If we saturate at the max here, then "next epoch" is not really valid anymore.

Wdyt about using checked add to panic on overflow?

File: cli/src/stake.rs [hunk 3]
```diff
@@ -2477,7 +2475,7 @@ pub fn get_epoch_boundary_timestamps(
                 break block_time;
             }
             Err(_) => {
-                epoch_start_slot += 1;
+                epoch_start_slot = epoch_start_slot.saturating_add(1);
```
Comment (brooksprumo) [inline cli/src/stake.rs:2478]: If we saturate here, will that create an infinite loop?

**Review by brooksprumo — COMMENTED** [review]
I didn't make it through everything yet, but but I've left a few comments.

In general, I'm not a fan of saturating math. I find I need to audit all the uses here, because sometimes there are subtle changes in behavior at the edges. I think we often believe "this can never overflow". So I tend to prefer checked math and then unwrap/expect/panic/return an error instead.

**Comment by illia-bobyr** [comment]
> I didn't make it through everything yet, but but I've left a few comments.
> 
> In general, I'm not a fan of saturating math. I find I need to audit all the uses here, because sometimes there are subtle changes in behavior at the edges. I think we often believe "this can never overflow". So I tend to prefer checked math and then unwrap/expect/panic/return an error instead.

Thank you for looking through it.
Sorry that I dumped such a big change as a single blob.
It is already split into commits, and I have incorrectly assumed it will help with the review %)

Let me publish individual commits as individual PRs.
Also, I do not disagree with the suggestion to use checked math.
My thinking with the saturating math is that is it almost always better than the wrapped math, and has the same performance.  So it is my default choice.
In a few cases, I did think about what exactly happens at the edge case, and saturation was pretty good.
But it is possible that checking and panicking could be a better user experience.

**Comment by illia-bobyr** [comment]
Let's work on this one commit at a time.
First change: https://github.com/anza-xyz/agave/pull/746

File: cli/src/test_utils.rs (hunk 2 above)
Comment (illia-bobyr) [reply cli/src/test_utils.rs:55]: Will do.

Comment (illia-bobyr) [reply cli/src/test_utils.rs:55]: This value is only used for the console message in the next statement.
And this function is only used in tests.
So, if this addition would actually saturate, it would only cause the message to be slightly incorrect.
I figured, it probably is not worth complicating this further with an explicit check.

File: cli/src/stake.rs (hunk 3 above)
Comment (illia-bobyr) [reply cli/src/stake.rs:2478]: Except for a test case that is checking for a boundary case, it would take a practically infinite amount of time to saturate a `u64` %)
But, I've changed this to a checked add.
This whole addition is actually incorrect in the first place.
Added a note.
